diff --git a/hrt/prefab/fx/Emitter.hx b/hrt/prefab/fx/Emitter.hx
index 2d108daf..966f5909 100644
--- a/hrt/prefab/fx/Emitter.hx
+++ b/hrt/prefab/fx/Emitter.hx

@@ -1084,11 +1084,12 @@ class EmitterObject extends h3d.scene.Object {
 					lockAxis.set(0, 0, 1);
 					frontAxis.set(0, 1, 0);
 			}
-
 			var lookAtPos = tmpVec;
 			lookAtPos.load(getScene().camera.pos);
+            lookAtPos.w = 1.0;
+
 			var invParent = parent.getInvPos();
-			lookAtPos.transform(invTransform);
+			lookAtPos.transform(invParent);
 			var deltaVec = new h3d.Vector(lookAtPos.x - x, lookAtPos.y - y, lookAtPos.z - z);
 
 			var invParentQ = tmpQuat;
			